app/editor/src/features/admin/topics/TopicList.tsx

```diff
@@ -59,40 +56,79 @@ export const TopicList: React.FC = () => {
+    const topic = items.find((i) => i.id === topicId);
+    if (!topic) return;
+    const removedTopic = { ...topic, isEnabled: false };
+    await api.updateTopic(removedTopic);
+    setItems(items.filter((t) => t.id !== topic.id));
+    setTopics(topics.map((item) => (item.id === topic.id ? removedTopic : item)));
+    if (!!id) navigate('/admin/topics');
+    toast.success(`${topic.name} has successfully been deleted.`);
+  };
+
   const handleSubmit = async (values: ITopicModel) => {
     try {
       var results = [...items];
+      let updatedTopics: ITopicModel[];
+      const existsTopic = topics.find((x) => x.name === values.name);
       if (values.id === 0) {
-        const topic = topics.find((x) => x.name === values.name);
-        if (!topic) {
+        if (!existsTopic) {
           const result = await api.addTopic(values);
           results = [...items, result];
+          updatedTopics = [...topics, result];
         } else {
-          const result = await api.updateTopic({
-            ...topic,
-            isEnabled: values.isEnabled,
-            description: values.description,
-            topicType: values.topicType,
-          });
-          if (topic.isEnabled) results = items.map((i) => (i.id === result.id ? result : i));
-          else if (values.isEnabled) results = [...items, result];
+          if (existsTopic.isEnabled) {
+            toast.warn(`${values.name} already exists.`);
+            return;
+          } else {
+            const result = await api.updateTopic({
+              ...existsTopic,
+              isEnabled: values.isEnabled,
+              topicType: values.topicType,
+            });
+            results = [...items, result];
+            updatedTopics = topics.map((i) => (i.id === existsTopic.id ? result : i));
+          }
         }
       } else {
-        const result = await api.updateTopic(values);
-        results = items.map((i) => (i.id === values.id ? result : i));
+        if (!existsTopic || values.id === existsTopic.id) {
+          const result = await api.updateTopic(values);
+          results = items.map((i) => (i.id === values.id ? result : i));
+          updatedTopics = topics.map((i) => (i.id === topic.id ? result : i));
+        } else {
+          toast.warn(`${values.name} already exists.`);
+          return;
+        }
       }
       setItems(results.filter((x) => x.isEnabled));
+      setTopics(updatedTopics);
       toast.success(`${values.name} has successfully been saved.`);
+      // Do we have a better way here for the clean-up after adding a new topic?
+      if (!id || id === '0') {
+        values.name = defaultTopic.name;
+        values.topicType = defaultTopic.topicType;
+        if (id === '0') navigate('/admin/topics');
+      }
     } catch {
       // Ignore error as it's handled globally.
     }
   };
 
+  const changeParentOverflow = () => {
+    let mainElement = document.querySelector('main');
```

Comment:
You should use a `useRef` instead of traveling the document tree.

Reply:
Code has been updated to be just a bit more simplified for now, and it will only be executed once in this case. 

As the main tag is located in a very top level, using useRef does it mean we need to pass the reference down across different layers? Currently do we have something similar implemented somewhere in the system or is there a simple way to achieve this? 

Thanks